iree/compiler/Dialect/HAL/Target/MetalSPIRV/SPIRVToMSL.cpp

```diff
@@ -37,6 +40,66 @@ class SPIRVToMSLCompiler : public SPIRV_CROSS_NAMESPACE::CompilerMSL {
     if (workgroupSize.constant != 0) return {0, 0, 0};
     return {workgroupSize.x, workgroupSize.y, workgroupSize.z};
   }
+
+  // A struct containing a resource descriptor's information.
+  struct Descriptor {
+    uint32_t set;
+    uint32_t binding;
+
+    Descriptor(uint32_t s, uint32_t b) : set(s), binding(b) {}
+
+    friend bool operator<(const Descriptor& l, const Descriptor& r) {
+      return std::tie(l.set, l.binding) < std::tie(r.set, r.binding);
+    }
+  };
+
+  // Returns all all resource buffer descriptors' set and binding number pairs
+  // in increasing order.
+  std::vector<Descriptor> getBufferSetBindingPairs() {
+    std::vector<Descriptor> descriptors;
+
+    // Iterate over all variables in the SPIR-V blob.
+    ir.for_each_typed_id<SPIRV_CROSS_NAMESPACE::SPIRVariable>(
+        [&](uint32_t id, SPIRV_CROSS_NAMESPACE::SPIRVariable& var) {
+          auto storage = var.storage;
+          switch (storage) {
+              // Non-interface variables. We don't care.
+            case spv::StorageClassFunction:
+            case spv::StorageClassPrivate:
+            case spv::StorageClassWorkgroup:
+              // Builtin variables. We don't care either.
+            case spv::StorageClassInput:
+              return;
+            default:
+              break;
+          }
+          if (storage == spv::StorageClassUniform ||
+              storage == spv::StorageClassStorageBuffer) {
+            uint32_t setNo = get_decoration(id, spv::DecorationDescriptorSet);
+            uint32_t bindingNo = get_decoration(id, spv::DecorationBinding);
+            descriptors.emplace_back(setNo, bindingNo);
+            return;
+          }
+          // TODO(antiagainst): push constant
+          llvm_unreachable("unspported storage class in SPIRVToMSLCompiler");
```

Comment:
Is this actually unreachable, or just unimplemented? What's the LLVM convention for unimplemented?

Reply:
Some of the cases are unimplemented. :) There is no `llvm_unimplemented` or something like `absl::UnimplementedError`. Using `llvm_unreachable` is sort of the convention by grabbing the codebase: https://github.com/llvm/llvm-project/search?q=unimplemented